### ISSUE TITLE
Refactoring observations about OpenSeadragonViewer.js, for fixing bugs and aiding maintainability

### ISSUE THREAD

**Issue by aeschylus** [ProjectMirador/mirador#2283]
We're already experiencing some OSD-related bugs (#2281, #2280, #2279, #2120 ), and adding features that pertain to annotation rendering have proven difficult and the solutions hard to follow/mimic for future contributors, as well as difficult for us to extend and build on. For maintainability and future development, I want to get to the root cause of some of these problems and create a clear seam between OSD and our app, with a pattern for how to interact with OSD from redux. These are some initial thoughts and questions to get the discussion started.

## Context 
The effectiveness of redux state management depends upon a functional, stateless rendering system. React provides this for most of our interface elements. In this paradigm, the rendering pipeline is typically:

**the full state is yielded from redux => react diffs the props => react makes the visual changes in elements that reference the props.**

OpenSeadragon is not a stateless rendering system like react, and we haven't been able to rely on this pattern to implement features. Therefore, currently, some state lives in openseadragon, some lives in redux, some lives in both places, some lives as properties on the component class, and some lives as code in the form of promises and sentinel statements - we are mutating state across paradigms, which is causing bugs and making it hard to reason about the implementation of new features. I would like to create a clear boundary between these paradigms, with patterns describing which things should be put on which sides of this boundary line, in such a way that it is easy to see where and how we would drop in another viewer. This mixing paradigms was a problem in Mirador 2 as well, so I want us to try to head it off at the pass if possible.

### What paradigms are we talking here?
There are 4 state-management paradigms attempting to coexist and interact here, and we need names for them.

1. **Immutable Rendering** - This is the system used in the react render function. This function is run every single time any that it depends on changes. It doesn't need to know the _type_ of change that occurred because it will render exactly what the data it receives tells it to render, and will always create the same output given the same input.
1. **OO API usage** - This is the system used in some places where we update openSeadragon (`fitBounds`, `addTileSource`). It lets openSeadragon decide how to update its internal state and what side effects to produce as a result of our method call. Calling the same method with the same inputs _will **NOT**_ produce the same result each time unless OSDs internal state is also the same when it is called. We therefore cannot guarantee the outcomes of this behaviour unless we know what the internal state is. There is another aspect of this paradigm which is using constructors to do initialisation and then accessors to make changes to properties. Because these aspects rely on the execution context of the object, in is sometimes hard to know when certain properties will or will not be defined, especially when object initialisations occur inside of react lifecycle methods.
1. **Inline Callbacks** - State is valid only within the closure of the written code, and can be different based on when and where the code is called because of `this` and lexical scope. Inline callbacks make it hard to reason about the effects they produce because they might be overridden by a different function elsewhere at a different execution time. Examples include the `addTileSource` callbacks and annotation rendering callbacks.
1. **Event-based Callback (inversion of control)** - OpenSeadragon provides a number of events that can be subscribed to, such as `update-viewport`, `zoom`, etc. This is a way to cede control of execution time to OSD and simply receive updates. This can be useful for rendering (for annotations) or for dispatching redux actions.
1. **Immediate-mode Rendering** - because it is based on HTML5 canvas, OpenSeadragon's overarching UI model is an "[immediate-mode-gui](https://docs.microsoft.com/en-us/windows/desktop/learnwin32/retained-mode-versus-immediate-mode)". This is a frame-based animation system often used in video games. It involves a loop that is run every 16 milliseconds that fully redraws the entire UI based on the current state of a "scene" (a big object or tree describing what needs to be drawn). When we directly set the internal properties of OSD by not using its API, we can potentially disrupt its diff checking, since we are essentially 

## Questions to Help Refine our Understanding
1. **How would we eliminate sentinels that check for the existence of loaded instances?**
https://github.com/ProjectMirador/mirador/blob/12825008581f65438f9026e061b2e6c2bc5f5bc7/src/components/OpenSeadragonViewer.js#L35-L37
https://github.com/ProjectMirador/mirador/blob/12825008581f65438f9026e061b2e6c2bc5f5bc7/src/components/OpenSeadragonViewer.js#L51-L54
https://github.com/ProjectMirador/mirador/blob/12825008581f65438f9026e061b2e6c2bc5f5bc7/src/components/OpenSeadragonViewer.js#L88-L98

Ideally, I think there should be no uncertainty about whether a viewer instance exists at a given step in the execution or what the state of the ref is. If there is uncertainty, it is likely because we have done something wrong, and need to find the root cause of the unpredictable behaviour.

2. **Why are #2120 and #2281 happening?** - in #2120, we see all the canvases re-render and re-zoom themselves with an animation. A similar thing happens in #2281. This is because the initialisation code is being run again, but with the viewer defined, so the "immediately" flag passed to the zoom functions is "false". What is a better name (and a more explicit way for us to identify) for this case than "the viewer instance is defined as a property on this component? How can we make this control flow more predictable?
2. **What should go in the initial setup of `componentDidMount`, and what should go in `componentDidUpdate`?**
3. **Which processes can we let OpenSeadragon own?** - OpenSeadragon provides an event system that communicates to the developer where it is in its render cycle. With annotations, for example, it is possible to update them whenever the `animation` event is fired (every 16ms, whenever there is any discrepancy between its internal "currently rendered state" and "target render state").
4. **When do we need to set the zoom or constraints? When does it need to be animated, and when does it need to be immediate?**  - I think if we can say this in English, we can have an easier time coding it up. Right now we implicitly tie this to Tilesource loading with Promises, but I don't think that is either necessary or wise. What redux-level actions should cause a redux=>OSD zoom instruction? How would these be represented in the Redux state?
5. **How could we eliminate promises from inside the component lifecycle methods?** - These Promises add multiple layers of asynchrony that are not accessible to the rest of the application and cannot be handled in a "round-trip manner". Their resolutions trigger method calls of the component's own methods at uncontrollable times, essentially hiding state in several paradigms.
https://github.com/ProjectMirador/mirador/blob/12825008581f65438f9026e061b2e6c2bc5f5bc7/src/components/OpenSeadragonViewer.js#L152-L164
https://github.com/ProjectMirador/mirador/blob/12825008581f65438f9026e061b2e6c2bc5f5bc7/src/components/OpenSeadragonViewer.js#L81-L86
https://github.com/ProjectMirador/mirador/blob/12825008581f65438f9026e061b2e6c2bc5f5bc7/src/components/OpenSeadragonViewer.js#L251-L253
7. **When is the viewer created, "closed", or destroyed?**
8. **What things should OSD tell our application?** - If we were to pretend that OSD were a react component, when would it call our redux actions? Which redux actions would it call, and can all of them be called using its provided [events](https://openseadragon.github.io/docs/OpenSeadragon.Viewer.html) (`update-viewport`, `tilesource-added`, etc.)
9. **What things should we tell OpenSeadragon?** - Assuming OSD could listen to the redux actions like a store (which it can't presently, we simply get an undifferentiated "all state is different now" update from connect in `componentDidUpdate`), which [action-types](https://github.com/ProjectMirador/mirador/blob/master/src/state/actions/action-types.js) would it respond to?
10. **Would it be easier to update OSD if we had an event system?** - As mentioned above and in the redux [docs](https://redux.js.org/api/store#subscribe), we do not receive typed state updates from the redux store. We only know in `componentDidUpdate` that some of the props originally passed are now different. We don't know which ones changed unless we manually diff them. This can be addressed by #2084, or perhaps by not relying on the react wrapper to control the update cycle. This was the strategy chosen by another Stanford project some years ago that also combined react+redux+OSD ([Chinese Graves Project](https://github.com/sul-cidr/graves/tree/master/app/assets/javascripts/read/src)).

## Existing Code
Mirador 2 was forced to directly confront many of these issues when we wanted to implement layers/multispectral image display. Several libraries were written to deal with this, and they still contain a great deal of useful code.

[**eventedCanvas**](https://github.com/sul-dlss/iiif-evented-canvas) represents a single canvas with all its image resources. It is an object that parses a IIIF canvas JSON node into a set of properties with methods to update them. **Writes are synchronous**, while an event emitter is accepted to provide **asynchronous reads**. The intent here was that OSD would only react to updates on this object through the event system. There are two implementations of this, one which is easier to read ([osdRenderer](https://github.com/sul-dlss/iiifManifestLayouts/blob/remove-incorrect-osd-references%23134/src/osdRenderer.js)), and one which was actually used in Mirador 2's image view ([Mirador event system bindings](https://github.com/ProjectMirador/mirador/blob/release-2.x/js/src/workspaces/window.js#L263-L315), [renderer initialisation](https://github.com/ProjectMirador/mirador/blob/release-2.x/js/src/widgets/imageView.js#L685-L712), [OSD reactions/callbacks](https://github.com/ProjectMirador/mirador/blob/release-2.x/js/src/widgets/imageView.js#L524-L608)).

[**iiifLayoutFunctions**](https://github.com/sul-dlss/iiif-layout-functions) ([demo](http://sul-dlss.github.io/iiif-layout-functions/example/)) was written to calculate layout parameters for the different ways canvases could be framed in the viewport for book, scroll, and single view.

[**iiifManifestLayouts**](https://github.com/sul-dlss/iiifManifestLayouts/tree/remove-incorrect-osd-references%23134), ([demo](http://sul-dlss.github.io/iiifManifestLayouts/example/)) unified these in an incomplete way, and attempted to be what in our codebase is the `<PrimaryWindow>` component, though it managed all of its own state and had its own event system. This overreached and was never completed, but some of one part of it (referenced above: osdRenderer) was ripped out and put into Mirador 2, creating the layers feature (also linked above).

The same strategies won't necessarily work in exactly the same way for us here because of our choice of framework, but they do provide clear seams that should help us find our own seams.

**Comment by mejackreed** [comment]
There is some great information contained here. While not exactly actionable for the M3 launch, going to close this issue.
